rpc/shared/shared/src/main/scala/io/udash/rpc/UdashRPCFramework.scala

```diff
@@ -138,69 +116,85 @@ trait UdashRPCFramework extends GetterRPCFramework with ProcedureRPCFramework wi
     }
   }
 
-  implicit val RPCResponseCodec = new GenCodec[RPCResponse] {
-    override def read(input: Input): RPCResponse = {
-      val obj = input.readObject()
-      var response: Any = null
-      var callId: String = null
-      var cause: String = null
-      var errorMsg: String = null
-      var tpe: String = null
-      while (obj.hasNext) {
-        val fi = obj.nextField()
-        fi.fieldName match {
-          case "response" =>
-            response = RawValueCodec.read(fi)
-          case "callId" =>
-            callId = fi.readString()
-          case "cause" =>
-            cause = fi.readString()
-          case "errorMsg" =>
-            errorMsg = fi.readString()
-          case "type" =>
-            tpe = fi.readString()
+  def RPCResponseCodec(exceptionsRegistry: ExceptionCodecRegistry): GenCodec[RPCResponse] =
+    new GenCodec[RPCResponse] {
+      override def read(input: Input): RPCResponse = {
+        val obj = input.readObject()
+        val tpe = obj.nextField().assertField("type").readString()
+
+        tpe match {
+          case "RPCResponseSuccess" =>
+            val response = RawValueCodec.read(obj.nextField().assertField("response"))
+            val callId = obj.nextField().assertField("callId").readString()
+            RPCResponseSuccess(response.asInstanceOf[RawValue], callId)
+          case "RPCResponseFailure" =>
+            val cause = obj.nextField().assertField("cause").readString()
+            val errorMsg = obj.nextField().assertField("errorMsg").readString()
+            val callId = obj.nextField().assertField("callId").readString()
+            RPCResponseFailure(cause, errorMsg, callId)
+          case "RPCResponseException" =>
+            val wrapper = obj.nextField().assertField("exception").readObject()
+            val exceptionName = wrapper.nextField().assertField("type").readString()
+            val exception = exceptionsRegistry.get[Throwable](exceptionName).read(wrapper.nextField().assertField("data"))
```

Comment:
Actually, if we serialize stack trace, we should do it for the whole cause chain... especially considering the fact that the last cause and its stack trace is usually the most interesting.
Possibly we should also serialize all the causes themselves - we can't assume excepton's codec is doing that. Your codecs for standard library exceptions aren't doing that (note: cause can be set on exception using `initCause`)